src/async/rpc_async.ml

```diff
@@ -1,166 +1,21 @@
-  result
-
-module GenServer () = struct
-  open Rpc
-
-  let funcs = Hashtbl.create 20
-  type implementation = server_implementation
-  let description : Idl.Interface.description option ref = ref None
-  let implement x = description := Some x; funcs
-
-  type ('a,'b) comp = ('a,'b) Result.result M.async
-  type rpcfn = Rpc.call -> Rpc.response Deferred.t
-  type funcs = (string, rpcfn option) Hashtbl.t
-  type 'a res = 'a -> unit
-
-  type _ fn =
-    | Function : 'a Param.t * 'b fn -> ('a -> 'b) fn
-    | Returning : ('a Param.t * 'b Idl.Error.t) -> ('a, 'b) M.t fn
-
-  let returning a b = Returning (a,b)
-  let (@->) = fun t f -> Function (t, f)
-
-  let rec has_named_args : type a. a fn -> bool =
-    function
-    | Function (t, f) -> begin
-        match t.Param.name with
-        | Some n -> true
-        | None -> has_named_args f
-      end
-    | Returning (t, e) ->
-      false
-
-  let declare : string -> string list -> 'a fn -> 'a res = fun name _ ty ->
-    let open Rresult.R in
-    (* We do not know the wire name yet as the description may still be unset *)
-    Hashtbl.add funcs name None;
-    fun impl ->
-      begin
-        (* Sanity check: ensure the description has been set before we declare
-           any RPCs *)
-        match !description with
-        | Some _ -> ()
-        | None -> raise NoDescription
-      end;
-      let rpcfn =
-        let has_named = has_named_args ty in
-        let rec inner : type a. a fn -> a -> call -> response Deferred.t = fun f impl call ->
-          match f with
-          | Function (t, f) -> begin
-              let is_opt = match t.Param.typedef.Rpc.Types.ty with | Rpc.Types.Option _ -> true | _ -> false in
-              let (arg_rpc, call') =
-                match get_arg call has_named t.Param.name is_opt with
-                | Result.Ok (x,y) -> (x,y)
-                | Result.Error (`Msg m) -> raise (MarshalError m)
-              in
-              let z = Rpcmarshal.unmarshal t.Param.typedef.Rpc.Types.ty arg_rpc in
-              match z with
-              | Result.Ok arg -> inner f (impl arg) call'
-              | Result.Error (`Msg m) -> raise (MarshalError m)
-            end
-          | Returning (t,e) -> begin
-              Deferred.bind impl.M.async (function
-                  | Result.Ok x -> Deferred.return (success (Rpcmarshal.marshal t.Param.typedef.Rpc.Types.ty x))
-                  | Result.Error y -> Deferred.return (failure (Rpcmarshal.marshal e.Idl.Error.def.Rpc.Types.ty y)))
-            end
-        in inner ty impl
-      in
-
-      Hashtbl.remove funcs name;
-      (* The wire name might be different from the name *)
-      let wire_name = get_wire_name !description name in
-      Hashtbl.add funcs wire_name (Some rpcfn)
-
+module AsyncIdl = Idl.Make(AsyncM)
+module M = struct
+  include AsyncIdl.ImplM
+  (* TODO: can we avoid Obj.magic here? *)
+  let deferred x : 'a Async.Deferred.t = Obj.magic (AsyncM.unbox x)
```

Comment:
Yeah, I think this was the line I didn't cross when trying to do this...! Not that I'm saying it's a showstopper because unifying these things certainly reduces a large amount of copy-pasted code.

Reply:
I think this can be avoided, I have only to figure out what is the best way